contributors/design-proposals/kubeadm-upgrades.md

```diff
@@ -0,0 +1,350 @@
+1. Performs post-upgrade tasks like applying the new kube-dns and kube-proxy manifests and version-specific upgrade steps like upgrading Bootstrap Tokens from alpha in v1.7 to beta in v1.8. 
+
+### Architectural upgrade implementation options
+
+**Upgrade code in the CLI tool vs controller running in-cluster:**
+
+On a high level, the upgrade could be implemented in two primary ways:
+
+1. As synchronous code executing client-side in the kubeadm CLI binary.
+1. As an asynchronous controller that handles the upgrade implementation; with the kubeadm CLI just invoking the upgrade
+
+The historical preference was to build a reusable Operator that does the upgrading for us.
+An important note here is that in *either way we would deliver the exact same 'kubeadm upgrade' UX.*
+
+A head-to-head breakdown of pros/cons for the two methods outlined:
+
+1. Build the upgrading code into the `kubeadm` CLI command
+ -  Pros:
+   - *Simplicity*; Easier and simpler to implement. A `kubeadm upgrade` command will be built in any case.
+   - The [kubeadm v1.6 -> v1.7 alpha upgrade method](#TODO) used this option successfully
+   - Being able to easily have an interactive flow where the user can inspect the changes that are about to be applied and accept/deny for instance.
+   - No need to deal with version skew between the Operator image tag/version and the kubeadm CLI version
+   - No need to define an API (CRD or the like) between the client and the Operator
+ - Cons:
+   - Maybe not as reusable as an Operator inside the cluster would be
+
+1. Build an Operator
+ - Pros:
+   - The idea is that it would be generic and could be re-used by others
+   - Other operators are already handling complex upgrades successfully, and we could even standardise patterns
+ - Cons:
+   - Would have to deal with version skew between the controller and the CLI
+   - Would have to deal with pre-pulling the image to a specific node in the case of no internet connection
+   - Would have to define an API (CRD or the like) between the client and the Operator (more time-consuming to get right)
+
+**Decision**: Keep the logic inside of the kubeadm CLI (option 1) for the implementation in v1.8.0.
+
+We *might* revisit this and build an upgrade operator in future, either before GA of this functionality or after, in v2.
+
+**Rewrite manifests completely from config object or mutate existing manifest:**
+
+Instead of generating new manifests from a versioned configuration object, we could try to add "filters" to the existing manifests and apply different filters depending on what the upgrade looks like. This approach, to modify existing manifests in various ways depending on the version bump has some pros (modularity, simple mutating functions), but the matrix of functions and different policies would grow just too big for this kind of system, so we voted against this alternative in favor for the solution above.
```

Comment:
did we actually vote? maybe say decided instead?

Reply:
right, we didn't actually vote. Thanks for the reword suggestion there.